### PR TITLE
feat: Make Jest unit tests run faster in GitHub actions

### LINKED ISSUE #25680: Make Jest unit tests run faster in GitHub actions
### What is this about?

Mentioned in https://github.com/MetaMask/metamask-extension/pull/25570, quoting @davidmurdoch :

> As discussed in slack, let's follow up this PR with an investigation into making it go faster (to match the speeds at which Circleci ran the same jobs).
> 
> But in case it takes a while to do that work: can you add a new GitHub Issue and a comment in this file linking to the issue, explaining that we should look into making it go faster?
> 
> I just don't want people copying this work in the (near) future, not noticing the slow down, and suddenly our tests take hours to pass CI. Hopefully a comment in here will help prevent that.

By migrating the jest unit tests from CircleCI to GitHub actions the time to run unit tests has increased to 15 minutes from 5 minutes. This can be for several reasons:

1. CircleCI and GitHub runners have different hardware
   - CircleCI: 3 CPU, 6 GB RAM VM
   - GitHub: 2 CPU, 7 GB RAM VM 
2. CircleCI runs the tests in parallel (parallelism: 8).
   - This would also be possible with GitHub actions, but we would need to restructure our testing code a bit. Refactor/remove `test/run-unit-tests.js`, as it was written with CircleCI in mind. We would need to make our testing command and workflow more `sharding` friendly.
   - Sources to implement GitHub actions sharding for jest: 
     - https://remarkablemark.org/blog/2022/05/13/jest-shard-test-github-actions/
     - https://medium.com/@mfreundlich1/speed-up-your-jest-tests-with-shards-776e9f02f637
     - https://didikmulyadi.medium.com/reducing-jest-execution-time-by-shard-parameters-and-github-matrix-from-90-120-minutes-to-5-7-ee86e91d7a99 
     - https://blacksmith.sh/blog/how-to-run-jest-tests-faster-in-github-actions

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

The unit tests run under 5 minutes, like it did on CircleCI

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by itsyoboieltr**
## **Description**

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25726?quickstart=1)

Based on https://github.com/MetaMask/metamask-extension/issues/25680, by migrating the jest unit tests from CircleCI to GitHub actions the time to run unit tests has increased to 15 minutes from 5 minutes. This PR makes Jest unit tests run faster in Github actions. This is achieved by some improvements to the testing infrastructure.

1. **Sharding**: the tests are run on 6 machines in parallel, instead of a single machine. This saves around ~8-9 minutes on average.
2. **Optimized coverage generation**: I removed the generation of html reports in CI, as they are only used for local debugging. This should also save some time, as unused files are not being generated on CI.
3. **Merged development tests**: The two small tests under `build/transforms` were merged into the main test configuration. This means that one less machine needs to be started (as the development tests would require their own machine to run, due to the sharding introduced in this PR), saving around approx. 2 minutes of compute time.
4. **Minor fixes**: Now the `run-unit-tests` workflow will also run on push to the `develop` and `master` branches.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/25680,

## **Manual testing steps**

1. Run CI and see that tests are running fast (between 5-7 mins on average)

## **Screenshots/Recordings**

Not Applicable

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -1,12 +1,7 @@
 module.exports = {
-  collectCoverageFrom: [
-    '<rootDir>/app/scripts/**/*.(js|ts|tsx)',
-    '<rootDir>/shared/**/*.(js|ts|tsx)',
-    '<rootDir>/ui/**/*.(js|ts|tsx)',
-  ],
   coverageDirectory: './coverage',
   coveragePathIgnorePatterns: ['.stories.*', '.snap'],
-  coverageReporters: ['html', 'json'],
+  coverageReporters: ['json'],
```
Comment (Gudahtt) [inline jest.config.js:4]: HTML reports are the sole method I use to review coverage. I use them at least. The JSON report isn't human readable, we can't really use that to review coverage.

File: jest.config.js [hunk 2]
```diff
@@ -1,12 +1,7 @@
 module.exports = {
-  collectCoverageFrom: [
```
Comment (Gudahtt) [inline jest.config.js:2]: Removing these will remove coverage of any untested files. It means we'll have no measurement of how many files are completely untouched by tests.

File: development/jest.config.js [hunk 3]
```diff
@@ -1,11 +0,0 @@
-module.exports = {
-  displayName: '/development',
-  collectCoverageFrom: ['<rootDir>/build/transforms/**/*.js'],
-  coverageDirectory: '../coverage',
-  coverageReporters: ['json'],
-  resetMocks: true,
-  restoreMocks: true,
-  testEnvironment: 'node',
```
Comment (Gudahtt) [inline development/jest.config.js:8]: It looks like you've changed these tests to run on jsdom instead of node by removing this. Is that correct, and was that intentional?

File: jest.config.js (hunk 1 above)
Comment (Gudahtt) [reply jest.config.js:4]: If we're not uploading it anywhere in CI, we can disable the HTML reporter there though.

File: .github/workflows/run-unit-tests.yml [hunk 4]
```diff
@@ -1,32 +1,52 @@
...
   pull_request:
     types: [opened,reopened,synchronize]
 
 jobs:
-  test-unit-jest:
+  test-unit:
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        shard: [1, 2, 3, 4, 5, 6, 7, 8]
```
Comment (HowardBraham) [inline .github/workflows/run-unit-tests.yml:14]: I would slightly reduce sharding.  With one shard, the compute time is approximately 28 minutes.  With 8 shards, the compute time is approximately 48 minutes.  Time spent in `Setup environment` is fairly high, so more sharding has diminishing returns.  Maybe try 6 shards?

File: jest.config.js (hunk 1 above)
Comment (itsyoboieltr) [reply jest.config.js:4]: @Gudahtt  If I see it correctly, they are not uploaded or used anywhere in CI, which is why I removed it. But I am not adamant about this at all, we can put it back if you think it is useful to always generate html reports. For your manual coverage reviews, you can still use the html flag to override the config defaults.

File: jest.config.js (hunk 2 above)
Comment (itsyoboieltr) [reply jest.config.js:2]: @Gudahtt This is true, but then again, all the shards will run coverage for all the tests (instead of only those tests that are part of the shard). I guess we just have to take the performance hit there, as jest does not seem to support coverage generation only for a specific shard. I'll re-add this field to make sure we also have coverage for untested files.

File: development/jest.config.js (hunk 3 above)
Comment (itsyoboieltr) [reply development/jest.config.js:8]: @Gudahtt This was intentional, as the tests also run successfully on jsdom, where the rest of our test suite is running. Is there any reason to keep it on node? If there is, then we could put the jest node environment comment on top of the two affected test files.

File: jest.config.js (hunk 1 above)
Comment (Gudahtt) [reply jest.config.js:4]: What I mean is that this should be the default when locally running unit tests. The HTML report is the only way to review coverage locally. If we don't have that, it's hardly worth running tests with coverage at all.

File: development/jest.config.js (hunk 3 above)
Comment (Gudahtt) [reply development/jest.config.js:8]: The code under test here is run as part of our build system, so the Node.js environment is a better test environment because it matches the environment we'd be running the code in.

We could certainly deal with this using an environment comment. The comment doesn't seem ideal because then it would not be applied to new test modules, even though we'd want any new tests to be run in a Node.js environment (because this directory is used for development/build scripts, which are always run in Node.js). But maybe that's not a huge concern, we don't typically test our development/build scripts.

File: jest.config.js (hunk 1 above)
Comment (itsyoboieltr) [reply jest.config.js]: Oh, I understand now. Thank you for the clarification. I propose the following to keep both the performance and development experience benefits:

```ts
process.env.CI ? ['json'] : ['html', 'json'],
```

Is this okay with you? 😄

File: jest.config.js (hunk 2 above)
Comment (itsyoboieltr) [reply jest.config.js]: fixed, so I am resolving this

File: jest.config.js (hunk 1 above)
Comment (Gudahtt) [reply jest.config.js]: Yep, makes sense!

Comment (itsyoboieltr) [reply jest.config.js]: resolving as fixed, thank you for the suggestion!

File: development/jest.config.js (hunk 3 above)
Comment (itsyoboieltr) [reply development/jest.config.js:8]: I added the environment comments for now. In the future, if the number of development tests were to exponentially increase, we could reconsider running it with a separate configuration. A higher number of tests would justify spinning up a new machine to run the tests.

File: .github/workflows/run-unit-tests.yml (hunk 4 above)
Comment (itsyoboieltr) [reply .github/workflows/run-unit-tests.yml]: 6 shards: total workflow duration: 7m 12s, total compute time: 35m
5 shards: total workflow duration: 7m 48s, total compute time: 34m

Comment (itsyoboieltr) [reply .github/workflows/run-unit-tests.yml]: @HowardBraham  I think we can stick to 6 shards, let me know what you think.

**Review by Gudahtt — APPROVED** [review]
LGTM!

File: development/build/transforms/remove-fenced-code.test.js [hunk 5]
```diff
@@ -36,6 +39,10 @@ describe('build/transforms/remove-fenced-code', () => {
       lintTransformedFileMock.mockImplementation(() => Promise.resolve());
     });
 
+    afterEach(() => {
```
Comment (danjm) [inline development/build/transforms/remove-fenced-code.test.js:42]: Why was this addition needed?

File: jest.config.js (hunk 1 above)
Comment (danjm) [reply jest.config.js]: :+1:

File: .github/workflows/run-unit-tests.yml (hunk 4 above)
Comment (danjm) [reply .github/workflows/run-unit-tests.yml]: :+1:

File: development/build/transforms/remove-fenced-code.test.js (hunk 5 above)
Comment (itsyoboieltr) [reply development/build/transforms/remove-fenced-code.test.js:42]: @danjm Because the previously used jest development config had resetMocks enabled. The main jest config does not have this enabled, so we manually have to reset the mocks. In an ideal world, instead of resetting the mocks manually, we would enable resetMocks in the jest config. However, that would take a considerable amount of time, as we would need to migrate the main test suite to support resetMocks.

**Review by danjm — APPROVED** [review]

**Review by HowardBraham — APPROVED** [review]

**Review by vthomas13 — APPROVED** [review]
